File: db-query-backend/src/llm/ollama.provider.ts
// src/llm/ollama.provider.ts

// import fetch from 'node-fetch';
import {
  LLMProvider,
  GenerateSQLInput,
  GenerateSQLOutput,
} from './llm.types';
import { buildNLtoSQLPrompt } from '../nl/prompt.builder';

export class OllamaProvider implements LLMProvider {
  private readonly baseUrl = 'http://localhost:11434';

  async generateSQL(input: GenerateSQLInput): Promise<GenerateSQLOutput> {
    const prompt = buildNLtoSQLPrompt(
      input.dbType,
      input.schema,
      input.question,
    );

    const res = await fetch(`${this.baseUrl}/api/generate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: 'qwen2.5:3b',
        prompt,
        stream: false,
        options: {
          num_predict: 48,
          temperature: 0,
          top_p: 0.9,
          repeat_penalty: 1.1,
        },
      }),
    });

    const data: any = await res.json();
    const text = data?.response?.trim();

    if (!text) {
      throw new Error('Ollama returned empty response');
    }

    return {
      sql: text,
      raw: JSON.stringify(data),
      provider: 'ollama',
    };
  }
}

File: db-query-backend/app/llm/ollama_provider.py
"""Port of llm/ollama.provider.ts"""
import httpx

from app.config import get_settings
from app.llm.base import LLMProvider, GenerateSQLInput, GenerateSQLOutput
from app.llm.prompt_builder import build_nl_to_sql_prompt

settings = get_settings()


class OllamaProvider(LLMProvider):
    async def generate_sql(self, input: GenerateSQLInput) -> GenerateSQLOutput:
        prompt = build_nl_to_sql_prompt(input.db_type, input.schema, input.question)

        async with httpx.AsyncClient(timeout=settings.llm_timeout_ms / 1000) as client:
            resp = await client.post(
                f"{settings.ollama_base_url}/api/generate",
                json={
                    "model": settings.ollama_model,
                    "prompt": prompt,
                    "stream": False,
                    "options": {
                        "num_predict": 48,
                        "temperature": 0,
                        "top_p": 0.9,
                        "repeat_penalty": 1.1,
                    },
                },
            )
            resp.raise_for_status()
            data = resp.json()

        raw = data.get("response", "")
        return GenerateSQLOutput(sql=raw.strip(), raw=raw, provider="ollama")
